sample-code/src/langchain-azure-openai.ts

```diff
@@ -1,56 +1,129 @@
     language: 'german',
     text: 'What is the capital of France?'
   });
 }
 
 /**
- * Embed 'Hello, world!' using the OpenAI ADA model.
- * @returns An embedding vector.
+ * Perform retrieval augmeneted generation with the chat and embedding LangChain clients.
+ * @returns The answer from GPT.
  */
-export async function embedQuery(): Promise<number[]> {
-  const client = new AzureOpenAiEmbeddingClient({
-    modelName: 'text-embedding-ada-002'
+export async function invokeRagChain(): Promise<string> {
+  const __dirname = path.dirname(fileURLToPath(import.meta.url));
+  const resourcePath = resolve(__dirname, '../resources/orchestration.md');
+
+  // Create a text loader and load the document
+  const loader = new TextLoader(resourcePath);
+  const docs = await loader.load();
+
+  // Create a text splitter and split the document
+  const textSplitter = new RecursiveCharacterTextSplitter({
+    chunkSize: 2000,
+    chunkOverlap: 200
   });
-  return client.embedQuery('Hello, world!');
-}
+  const splits = await textSplitter.splitDocuments(docs);
 
-/**
- * Embed 'Hello, world!' and 'Goodbye, world!' using the OpenAI ADA model.
- * @returns An array of embedding vectors.
- */
-export async function embedDocument(): Promise<number[][]> {
-  const client = new AzureOpenAiEmbeddingClient({
-    modelName: 'text-embedding-ada-002'
+  // Initialize the embedding client with 0 retries for fast testing
+  const embeddingClient = new AzureOpenAiEmbeddingClient({
+    modelName: 'text-embedding-ada-002',
+    maxRetries: 0
+  });
+
+  // Create a vector store from the document
+  const vectorStore = await MemoryVectorStore.fromDocuments(
+    splits,
+    embeddingClient
+  );
+
+  // Create a prompt template
+  const promptTemplate = ChatPromptTemplate.fromTemplate(
+    `You are an assistant for question-answering tasks.
+      Use the following pieces of retrieved context to answer the question.
+      If you don't know the answer, just say that you don't know.
+      Use ten sentences maximum and keep the answer concise.
+      Also prioritize code examples.
+
+      Question: {question}
+      Context: {context}
+      Answer:`
+  );
+
+  // Initialize the chat client with 0 retries for fast testing
+  const llm = new AzureOpenAiChatClient({
+    modelName: 'gpt-35-turbo',
+    maxRetries: 0
+  });
+
+  // Create a chain to combine an LLM call with the prompt template and output parser
+  const ragChain = await createStuffDocumentsChain({
+    llm,
+    prompt: promptTemplate,
+    outputParser: new StringOutputParser()
+  });
+
+  // Create a retriever for the vector store
+  const retriever = vectorStore.asRetriever();
+
+  // Create a prompt
+  const prompt = 'How do you use templating in the SAP Orchestration client?';
+
+  // Invoke the chat client combined with the prompt, prompt template, output parser and vector store context
+  return ragChain.invoke({
+    question: prompt,
+    context: await retriever.invoke(prompt)
```

Comment:
langchain doesn't automate this step? calling invoke inside of invoke looks a bit cursed..